model/DataBaseAccess.php

```diff
@@ -329,6 +395,70 @@ public function removeMultiplePermissions(array $data)
         }
     }
 
+    /**
+     * Remove batch permissions
+     *
+     * @access public
+     * @param array $data
+     * @return void
+     */
+    public function removeMultiplePermissionsNew(array $resources)
+    {
+        $groupedRemove = [];
+        $eventsData = [];
+
+        foreach ($resources as $resource) {
+            foreach ($resource['permissions']['remove'] as $userId => $removePermissions) {
+                if (empty($removePermissions)) {
+                    continue;
+                }
+
+                $idString = implode($removePermissions);
+
+                $groupedRemove[$userId][$idString]['resources'][] = $resource['id'];
+                $groupedRemove[$userId][$idString]['privileges'] = $removePermissions;
+
+                $eventsData[] = [
+                    'userId' => $userId,
+                    'resourceId' => $resource['id'],
+                    'privileges' => $removePermissions,
+                ];
+            }
+        }
+
+        foreach ($groupedRemove as $userId => $priviligeGroups) {
+            foreach ($priviligeGroups as $privilegeGroup) {
+                $inQueryPrivilege = implode(',', array_fill(0, count($privilegeGroup['privileges']), ' ? '));
+                $inQueryResources = implode(',', array_fill(0, count($privilegeGroup['resources']), ' ? '));
+                $query = sprintf(
+                    'DELETE FROM %s WHERE %s IN (%s) AND %s IN (%s) AND %s = ?',
+                    self::TABLE_PRIVILEGES_NAME,
+                    self::COLUMN_RESOURCE_ID,
+                    $inQueryResources,
+                    self::COLUMN_PRIVILEGE,
+                    $inQueryPrivilege,
+                    self::COLUMN_USER_ID
+                );
+
+                $params = array_merge(
+                    array_values($privilegeGroup['resources']),
+                    array_values($privilegeGroup['privileges']),
+                    [$userId]
+                );
+
+                $this->getPersistence()->exec($query, $params);
+            }
+        }
+
+        foreach ($eventsData as $eventData) {
+            $this->getEventManager()->trigger(new DacRemovedEvent(
```

Comment:
Is the side effect of this syncronous event affecting queue performance

Reply:
Changed the code to use new Event